Streamline logic related to accessor derivation in MethodSynthesis & Namers

adriaanm opened the pull request:
Distinguish abstract ValDef from ValDef that needs a setter. 
(Later, all trait vals will also need a setter -- right now, it's only mutable ones.)

Similarly, make explicit whether a field is needed. (Later, fields will not be emitted for traits.)

Simplify decision whether to derive accessors. Comments in the original code below (modulo renaming & reduction of double negation in previous commit) explain why I believe my changes make sense:

```
def deriveAccessors(vd: ValDef) = vd.mods.isLazy || !(
     !owner.isClass
  || (vd.mods.isPrivateLocal && !vd.mods.isCaseAccessor) // this is an error -- now checking first
  || (vd.name startsWith nme.OUTER)
  || (context.unit.isJava) // pulled out to caller
  || isEnumConstant(vd)
)

def deriveAccessorTrees(vd: ValDef) = !(
     (vd.mods.isPrivateLocal && !vd.mods.isLazy) // lazy was pulled out to outer disjunction
  || vd.symbol.isModuleVar // pulled out to caller
  || isEnumConstant(vd))
```

These conditions are now captured by one method.

Review by @lrytz, @retronym & @SethTisue. I'll check bytecode diffs (should be empty) later tonight.

adriaanm committed:
Simplify decision whether to derive accessors

Originally (modulo renaming & reduction of double negation in previous commit):

```
def deriveAccessors(vd: ValDef) = vd.mods.isLazy || !(
     !owner.isClass
  || (vd.mods.isPrivateLocal && !vd.mods.isCaseAccessor) // this is an error -- now checking first
  || (vd.name startsWith nme.OUTER)
  || (context.unit.isJava) // pulled out to caller
  || isEnumConstant(vd)
)

def deriveAccessorTrees(vd: ValDef) = !(
     (vd.mods.isPrivateLocal && !vd.mods.isLazy) // lazy was pulled out to outer disjunction
  || vd.symbol.isModuleVar // pulled out to caller
  || isEnumConstant(vd))
```

With changes in comments above, these conditions are now captured by one method.

adriaanm commented:
This is in preparation of https://github.com/adriaanm/scala/tree/traits-late-fields, which will ultimately enable the interfaces-with-default-methods encoding of traits.

retronym commented on src/compiler/scala/tools/nsc/typechecker/MethodSynthesis.scala:
```diff
@@ -134,28 +134,26 @@ trait MethodSynthesis {
-      tree.symbol = (
-        if (mods.isLazy) {
+    // TODO: why is all this symbol creation disconnected from addDerivedTrees,
+    // which creates another list of Field/Getter/Setter factories???
```
Looks like this part enters the symbols (which needs to be done before typechecking), and `addDerivedTrees` is called to create the corresponding trees during typechecking.

I agree it would be cleaner to keep track of the factories we create here (maybe as a attachment on the ValDef's symbol?) and avoid the duplicated logic in `allValDefDerived`.

adriaanm replied:
- [x] I'll include your comment to improve on my snarky rhetoric.

retronym commented on src/compiler/scala/tools/nsc/typechecker/MethodSynthesis.scala:
```diff
@@ -305,17 +297,26 @@ trait MethodSynthesis {
+      //     performs its side effect according to lazy/strict semantics, but doesn't need to store its value
+      //     each access will "evaluate" the RHS (a literal) again
+      // We would like to avoid emitting unneccessary fields, but the required knowledge isn't available until after typer.
+      // The only way to avoid emitting & suppressing, is to not emit at all until we are sure to need the field, as dotty does.
```
> but the required knowledge isn't available until after typer.

Is it available _during_ typer, at the point that we are adding the derived field to a template? If so, at that juncture, we could omit the field and unlink the speculatively created field symbol from the class scope.

Given the initial creation of the derived field is already split between namer (`enterGetterSetter`) and typer (

adriaanm replied:
It is available during typer, but I'm not positive we are safe from introducing spurious cycles if we do this then. The cleanest approach IMO is to follow dotty, and not emit any fields until much later in the pipeline -- they are an implementation detail after all.

retronym commented on src/compiler/scala/tools/nsc/typechecker/MethodSynthesis.scala:
```diff
@@ -341,10 +343,13 @@ trait MethodSynthesis {
       }
-      private def setterRhs = (
-        if (mods.isDeferred || derivedSym.isOverloaded) EmptyTree
+      // TODO: when is `derivedSym.isOverloaded`??? is it always an error?
```
I don't see how we'd get here. In erroneous code like `class C(var x: Any) { var x: Any }` we issue the double def error during namer, and don't enter the second symbol in scope. So when this code is called (during typer), we don't end up with an overloaded setter.

Furthermore, we can't inherit overloads, as `DerivedSetter#derivedSym` uses `getterIn(enclClass)`, which only looks at `decls`.

I tried turning it into an assertion, but didn't trip it in `partest --neg --pos`.

adriaanm replied:
- [x] let's add the assert and ask the community build

retronym commented on src/compiler/scala/tools/nsc/typechecker/MethodSynthesis.scala:
```diff
@@ -341,10 +343,13 @@ trait MethodSynthesis {
-        if (mods.isDeferred || derivedSym.isOverloaded) EmptyTree
+      // TODO: when is `derivedSym.isOverloaded`??? is it always an error?
+      // `noFieldHere` does not imply DEFERRED (ask isDeferred if you must know)
+      // we automatically implement trait setters (where !isDeferred) in the subclass
```
- [x] This comment makes sense describing this patch, but I think it doesn't make much sense when reading the resulting code.

adriaanm replied:
agreed

adriaanm commented:
Progress? EDIT: nope, bug in my refactoring (in `doNotDeriveField`).

```
Testcase: testDefWithLazyVal2 took 0.096 sec
    FAILED
using toolbox typechecker
 expected:<... {
  lazy val test: []Unit = {
    scala.P...> but was:<... {
  lazy val test: [scala.]Unit = {
    scala.P...>
junit.framework.AssertionFailedError: using toolbox typechecker
 expected:<... {
  lazy val test: []Unit = {
    scala.P...> but was:<... {
  lazy val test: [scala.]Unit = {
    scala.P...>
    at scala.reflect.internal.PrinterHelper$.assertResultCode(PrintersTest.scala:53)
    at scala.reflect.internal.PrinterHelper$.assertPrintedCode(PrintersTest.scala:67)
    at scala.reflect.internal.ValAndDefPrintTests$class.testDefWithLazyVal2(PrintersTest.scala:1013)
    at scala.reflect.internal.PrintersTest.testDefWithLazyVal2(PrintersTest.scala:12)
```

adriaanm committed:
Review feedback from retronym

adriaanm commented:
I rebased, addressing some feedback in earlier commits, adding the assert last for easy reversal.

adriaanm commented:
/synch

adriaanm committed:
Streamline MethodSynthesis & Namers

Give Getter control over whether a setter is needed. For now,
only mutable ValDefs entail setters. In the new trait encoding,
a trait val will also receive a setter from the start.

Similarly, distinguish whether to derive a field from deferredness of the val.
(Later, fields will not be emitted for traits, deferred or not.)

adriaanm commented:
Running community build on merge of #4723 onto this one: https://scala-ci.typesafe.com/view/scala-2.12.x/job/scala-2.12.x-integrate-community-build/153/console

retronym commented:
LGTM
